### PR TITLE
IXLCell.TryGetValue improvements

### LINKED ISSUE #1469: Problem with reading cell value which has formula
## Read and complete the full issue template

Do not randomly delete sections. They are here for a reason.

**Do you want to request a *feature* or report a *bug*?**
- [x] Bug
- [] Feature
- [x] Question

**Did you test against the latest CI build?**
- [x] Yes
- [ ] No

If you answered `No`, please test with [the latest development build](https://ci.appveyor.com/project/ClosedXML/ClosedXML/branch/develop/artifacts) first.

**Version of ClosedXML**
0.95.3

e.g. 0.95.3

**What is the current behavior?**
If cell contains formula e.g. =Today()+2 etc. then we get the cell value as integer not the expected value.  

**What is the expected behavior or new feature?**
If the cell contains the date formula e.g. =Today()+2 then we expect to get 07/04/2020 as cell value.
I have question here as how I get the value of the cell after calculating the formula set on it.   
Complete this.

**Is this a regression from the previous version?**
No. Older version also not giving proper value. 

Regressions get higher priority. Test against the latest build of the previous minor version. For example, if you experience a problem on v0.95.3, check whether it the problem occurred in v0.94.2 too. 

## Reproducibility
**This is an important section. Read it carefully. Failure to do so will cause a 'RTFM' comment.**

Without a code sample, it is unlikely that your issue will get attention. Don't be lazy. Do the effort and assist the developers to reproduce your problem. Code samples should be [minimal complete and verifiable](https://stackoverflow.com/help/mcve). Sample spreadsheets should be attached whenever applicable. Remove sensitive information.

**Code to reproduce problem:**
```c#
public void Main()
{
    string fileName = @"<File Name>";
    var xlworkBook = new XLWorkbook(fileName);
    if (xlworkBook != null)
    {
               var sheet1 = xlworkBook.Worksheet("Sheet1");
               if (sheet1 != null)
                {
                    var rows = sheet1.RangeUsed().RowsUsed().Skip(1);
                    foreach (var row in rows)
                    {
                    var rowCell = row.Cell(1);
                    var rowString = rowCell.GetDateTime();// This will throw error if cell contains formula e.g. =Today()+2 etc. 
                    }
                }
            }
}
```

- [x] I attached a sample spreadsheet.  (You can drag files on to this issue)
[DateFormats.xlsx](https://github.com/ClosedXML/ClosedXML/files/4863569/DateFormats.xlsx)

### PR DESCRIPTION
**Pull request by igitur**
Fixes #1469 

Make `TryGetValue` try harder to coerce between data types.

### CODE REVIEW

File: ClosedXML/Excel/Cells/XLCell.cs [hunk 1]
```diff
@@ -1755,16 +1755,34 @@ private static bool TryGetDateTimeValue<T>(out T value, object currentValue)
...
+            if (currentValue.IsNumber())
             {
-                value = default;
-                return false;
+                var dbl1 = Convert.ToDouble(currentValue);
+                if (dbl1.IsValidOADateNumber())
+                {
+                    value = (T)Convert.ChangeType(DateTime.FromOADate(dbl1), typeof(T));
+                    return true;
+            }
```
Comment (Pankraty) [inline ClosedXML/Excel/Cells/XLCell.cs:1767]: The indentation is screwed up here, probably during the rebase.

File: ClosedXML/Excel/Cells/XLCell.cs [hunk 2]
```diff
@@ -1755,16 +1755,34 @@ private static bool TryGetDateTimeValue<T>(out T value, object currentValue)
...
+            {
             value = (T)Convert.ChangeType(ts, typeof(T));
             return true;
         }
 
+            if (Double.TryParse(currentValue.ObjectToInvariantString(), out Double dbl2) && dbl2.IsValidOADateNumber())
+            {
+                value = (T)Convert.ChangeType(DateTime.FromOADate(dbl2), typeof(T));
+                return true;
+            }
```
Comment (Pankraty) [inline ClosedXML/Excel/Cells/XLCell.cs:1780]: This block is not covered by tests; there might be something like this:
```
[Test]
public void TryGetValue_DateTime_Good3()
{
    IXLWorksheet ws = new XLWorkbook().Worksheets.Add("Sheet1");
    bool success = ws.Cell("A1").SetFormulaA1("=\"44\"&\"020\"").TryGetValue(out DateTime outValue);
    Assert.IsTrue(success);
    Assert.AreEqual(new DateTime(2020, 07, 08), outValue);
}
```
But I am not 100% sure this should work this way. Excel, for example, leaves text to be text, even if a cell type is Date

image

Comment (igitur) [reply ClosedXML/Excel/Cells/XLCell.cs:1780]: But this does work in Excel:
image

We could possibly move the string to Date conversion into the calculation engine, but I don't really see a down-side to leaving it here. If the user calls `cell.GetValue<DateTime>()` then I assume he really wants to get out a DateTime as much as possible.

Comment (igitur) [reply ClosedXML/Excel/Cells/XLCell.cs]: After thinking about it over the weekend, I now agree that we shouldn't do a "double-hop" to coerce values from one type to another. Converting a string to a number and then a number to a date, would be a "double-hop". This is only for the `TryGetValue<>()` cases. In the calc engine, coercing can be a bit more lenient, as is already the case. See updated unit test: `TryGetValue_DateTime_BadButFormulaGood`

**Pull request merged**